ntoskrnl/io/iomgr/deviface.c

```diff
@@ -412,6 +632,182 @@ IopOpenInterfaceKey(IN CONST GUID *InterfaceClassGuid,
+ * Must be freed with RtlFreeUnicodeString after using.
+ *
+ * @return NTSTATUS values in case of errors, STATUS_SUCCESS otherwise.
+ *
+ * @remarks Must be called at IRQL = PASSIVE_LEVEL in the context of a system thread
+ *
+ */
+NTSTATUS
+NTAPI
+IoGetDeviceInterfaceAlias(
+    _In_ PUNICODE_STRING SymbolicLinkName,
+    _In_ CONST GUID *AliasInterfaceClassGuid,
+    _Out_ PUNICODE_STRING AliasSymbolicLinkName)
+{
+    UNICODE_STRING AliasSymbolicLink;
+    UNICODE_STRING PrefixString;
+    UNICODE_STRING AliasGuidString;
+    UNICODE_STRING DeviceString, ReferenceString;
+    PKEY_VALUE_FULL_INFORMATION kvInfo;
+    HANDLE DeviceKey, AliasInstanceKey;
+    PVOID Buffer;
+    USHORT i;
+    NTSTATUS Status;
+
+    RtlInitEmptyUnicodeString(&PrefixString, NULL, 0);
+    RtlInitEmptyUnicodeString(&DeviceString, NULL, 0);
+    RtlInitEmptyUnicodeString(&ReferenceString, NULL, 0);
+    RtlInitEmptyUnicodeString(&AliasGuidString, NULL, 0);
+    RtlInitEmptyUnicodeString(&AliasSymbolicLink, NULL, 0);
+
+    DPRINT("IoGetDeviceInterfaceAlias() : symbolic link '%wZ'\n", SymbolicLinkName);
+
+    /* Sanity check */
+    if (SymbolicLinkName == NULL || AliasInterfaceClassGuid == NULL)
+    {
+        DPRINT1("IoGetDeviceInterfaceAlias() : invalid symbolic link or alias class GUID\n");
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &AliasGuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        goto Quit;
+    }
+
+    DPRINT("Alias GUID is '%wZ'\n", &AliasGuidString);
+
+    /* Get the device instance string of existing symbolic link */
+    Status = OpenRegistryHandlesFromSymbolicLink(SymbolicLinkName,
+                                                 KEY_QUERY_VALUE,
+                                                 NULL,
+                                                 &DeviceKey,
+                                                 NULL);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed to open device instance key for '%wZ' with status 0x%08lx\n", SymbolicLinkName, Status);
+        goto Quit;
+    }
+
+    Status = IopGetRegistryValue(DeviceKey, L"DeviceInstance", &kvInfo);
+    ZwClose(DeviceKey);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("Failed get device instance value with status 0x%08lx\n", Status);
+        goto Quit;
+    }
+
+    if (kvInfo->Type != REG_SZ || kvInfo->DataLength == 0)
+    {
+        DPRINT1("Wrong or empty instance value\n");
+        Status = STATUS_INVALID_PARAMETER;
+        goto Quit;
+    }
+
+    /* Convert received data to unicode string */
+    Buffer = (PVOID)((ULONG_PTR)kvInfo + kvInfo->DataOffset);
+    PnpRegSzToString(Buffer, kvInfo->DataLength, &DeviceString.Length);
+    DeviceString.MaximumLength = (USHORT)kvInfo->DataLength;
```

Comment:
It may be of interest to do a `(kvInfo->DataLength > MAXUSHORT)` check-and-fail just above (in the existing if-test).
So that when we get to set up this DeviceString here, we are 100% sure the string will be valid.
Also, you could maybe do this: `DeviceString.MaximumLength = DeviceString.Length`.
It doesn't seem that the rest of the code below depend on a NUL-terminator being present in this DeviceString.

Reply:
Added additional checks as suggested. I also think assigning `DeviceString.Length` to `DeviceString.MaximumLength` should work the same.